packages/fxa-auth-server/lib/payments/capability.ts

```diff
@@ -229,6 +245,117 @@ export class CapabilityService {
+      },
+      {}
+    );
+
+    const targetPlan = allPlansByPlanId[targetPlanId];
+
+    if (!targetPlan) throw error.unknownSubscriptionPlan(targetPlanId);
+
+    const { productSet: targetProductSet } = productUpgradeFromProductConfig(
+      targetPlan,
+      useFirestoreProductConfigs
+    );
+
+    if (!targetProductSet) {
+      return SubscriptionEligibilityResult.INVALID;
+    }
+
+    // Fetch all user's subscriptions from all sources
+    const [stripeSubscriptions, appleIapSubscriptions, playIapSubscriptions] =
+      await Promise.all([
+        this.fetchSubscribedPricesFromStripe(uid),
+        this.fetchSubscribedPricesFromAppStore(uid),
+        this.fetchSubscribedPricesFromPlay(uid),
+      ]);
+
+    const mapSubscriptions = (planIds: string[]) => {
+      return planIds
+        .map((planId) => allPlansByPlanId[planId])
+        .filter((plan) => plan);
+    };
+
+    const stripeSubscribedPlans = mapSubscriptions(stripeSubscriptions);
+    const appleIapSubscribedPlans = mapSubscriptions(appleIapSubscriptions);
+    const playIapSubscribedPlans = mapSubscriptions(playIapSubscriptions);
+
+    const iapSubscribedPlans = [
+      ...appleIapSubscribedPlans,
+      ...playIapSubscribedPlans,
+    ];
+
+    // Lookup whether user holds an IAP subscription with a shared product set to the target
+    const iapRoadblock = iapSubscribedPlans.some((abbrevPlan) => {
+      const { productSet } = productUpgradeFromProductConfig(
+        abbrevPlan,
+        useFirestoreProductConfigs
+      );
+
+      return productSet?.some((name) => targetProductSet.includes(name));
+    });
+
+    // Users with an IAP subscription to the productSet that we're trying to subscribe
+    // to should not be allowed to proceed
+    if (iapRoadblock) {
+      return SubscriptionEligibilityResult.BLOCKED_IAP;
+    }
+
+    const isSubscribedToProductSet = stripeSubscribedPlans.some(
+      (abbrevPlan) => {
+        const { productSet } = productUpgradeFromProductConfig(
+          abbrevPlan,
+          useFirestoreProductConfigs
+        );
+
+        return productSet?.some((name) => targetProductSet.includes(name));
+      }
+    );
+
+    if (!isSubscribedToProductSet) {
+      return SubscriptionEligibilityResult.CREATE;
+    }
+
+    // Use the upgradeEligibility helper to check if any of our existing plans are
+    // elegible for an upgrade and if so the user can upgrade that existing plan to the desired plan
+    for (const abbrevPlan of stripeSubscribedPlans) {
+      const eligibility = getSubscriptionUpdateEligibility(
+        abbrevPlan,
+        targetPlan,
+        useFirestoreProductConfigs
+      );
+
+      if (eligibility === SubscriptionUpdateEligibility.UPGRADE) {
+        return SubscriptionEligibilityResult.UPGRADE;
+      }
+
+      if (eligibility === SubscriptionUpdateEligibility.DOWNGRADE) {
+        return SubscriptionEligibilityResult.DOWNGRADE;
+      }
+    }
+
+    return SubscriptionEligibilityResult.INVALID;
```

Comment:
Happy to break this down farther. I would prefer to see this exist in it's own file but I was told `capability.ts` was the place this should go. I don't know that any individual parts of this would benefit from existing on their own, but breaking it up would still improve readability.

What are your thoughts WRT where this should live? Is my desire to move it out of this file completely reasonable?